src/main/java/city/parking/services/ParkingProcessService.java

```diff
@@ -0,0 +1,118 @@
+import java.time.LocalDateTime;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static city.parking.services.ParkingCostCalculationUtil.calculatePrimaryParkingCost;
+
+@Service
+public class ParkingProcessService {
+
+    private final ParkingProcessRepository parkingProcessRepository;
+
+    public ParkingProcessService(ParkingProcessRepository parkingProcessRepository) {
+        this.parkingProcessRepository = parkingProcessRepository;
+    }
+
+    public List<ParkingProcess> findAll() {
+        return parkingProcessRepository.findAll();
+    }
+
+    public List<ParkingProcess> findParkingProcessesByStage(ParkingProcess.Stage processStage) {
+        return parkingProcessRepository.findByStage(processStage);
+    }
+
+    public Optional<ParkingProcess> findById(Integer processId) {
+        return parkingProcessRepository.findById(processId);
+    }
+
+    public ParkingProcess startParkingProcess(ParkingProcess parkingProcess) {
+        Integer toBeStartedMeterId = parkingProcess.getMeterId();
+        ensureParkingMeterIsNotUsed(toBeStartedMeterId);
+
+        parkingProcess.setParkingStartTime(LocalDateTime.now());
+        parkingProcess.setStage(ParkingProcess.Stage.ONGOING);
+        parkingProcess.setPrimaryCurrencyCost(new Money(ParkingCostCalculationUtil.getPrimaryCurrency(), new BigDecimal("0")));
+        parkingProcessRepository.save(parkingProcess);
+        return parkingProcess;
+    }
+
+    public ParkingProcess updateParkingProcessStage(Integer processId,
+                                                    ParkingProcessPartialUpdateRequest parkingProcessPartialUpdateRequest) {
+        Optional<ParkingProcess> processOptional = parkingProcessRepository.findById(processId);
+        if (processOptional.isPresent()) {
+            ParkingProcess process = processOptional.get();
+            if (parkingProcessPartialUpdateRequest.isParkingMeterToBeStopped()) {
+                ensureParkingMeterIsRunning(process);
+                stopParkingMeter(process);
+            }
+            return process;
+        } else {
+            throw new ParkingProcessNotFoundException(processId);
+        }
+    }
+
+    public Set<Money> getParkingCosts(Integer processId) {
+        Optional<ParkingProcess> parkingProcessOptional = parkingProcessRepository.findById(processId);
+        if (parkingProcessOptional.isPresent()) {
+            Money primaryCurrencyCost = parkingProcessOptional.get().getPrimaryCurrencyCost();
+            return getDisplayableCosts(primaryCurrencyCost);
+        } else {
+            throw new ParkingProcessNotFoundException(processId);
+        }
+    }
+
+    private void ensureParkingMeterIsNotUsed(Integer toBeStartedMeterId){
+        //if parking meter is running or it is unpaid (parking process ONGOING or STOPPED_UNPAID stage)
+        //it cannot be started
+        List<ParkingProcess> ongoingParkingProcessesWithGivenMeterId =
```

Comment:
Ta logika wydaje się być niejasna dla mnie. Czy nie wystarczyłoby dla wszystkich procesów o danym id sprawdzić czy mają odpowiedni status? Niepotrzebnie są dwa zapytania do bazy, które (jak to sobie wyobrażam) korzystałyby z tego samego indeksu.

Reply:
Tu założyłam, że wszystkich procesów o danym id parkometru będzie z czasem bardzo dużo, stąd próba optymalizacji z wyciąganiem tylko procesów o danym id. Ale właśnie, żeby było to faktycznie szybsze rozwiązanie, niezbędny jest indeks na polu z etapem procesu parkowania (który generalnie byłyby przydatny w tej aplikacji)

Reply:
A mogłabyś zaproponować indeksy tutaj? Jakie byłyby potrzebne?

Reply:
Po zastanowieniu stwierdzam, że indeksowanie po polu stage nie jest dobrym pomysłem- nie będzie szybsze niż po prostu wyszukanie procesów z danym meterId i sprawdzenie ich statusu